File: python/day11.py
OFFS = [[-1, -1], [0, -1], [1, -1], [1, 0], [1, 1], [0, 1], [-1, 1], [-1, 0]]


def get_neighbor_counts(seats):
    h, w = len(seats), len(seats[0])
    res = [[0]*w for i in range(h)]
    for y in range(h):
        for x in range(w):
            n = 0
            for dx, dy in OFFS:
                x1, y1 = x + dx, y + dy
                n += x1 >= 0 and x1 < w and y1 >= 0 and y1 < h \
                    and seats[y1][x1] == '#'
            res[y][x] = n
    return res


def get_visible_neighbor_counts(seats):
    h, w = len(seats), len(seats[0])
    cansee = [[0]*w for i in range(h)]
    res = [[0]*w for i in range(h)]

    for dx, dy in OFFS:
        for j in range(h):
            for i in range(w):
                x = i if dx <= 0 else w - i - 1
                y = j if dy <= 0 else h - j - 1
                x1, y1 = x + dx, y + dy
                if x1 >= 0 and x1 < w and y1 >= 0 and y1 < h:
                    s = seats[y1][x1]
                    if s == '.':
                        cansee[y][x] = cansee[y1][x1]
                    elif s == '#':
                        cansee[y][x] = True
                    elif s == 'L':
                        cansee[y][x] = False
                else:
                    cansee[y][x] = False
                res[y][x] += cansee[y][x]
    return res


def step(seats, new_seats, neighbors_fn, max_occupied):
    neighbor_counts = neighbors_fn(seats)

    num_changed = 0
    total_occupied = 0
    for y in range(len(seats)):
        for x in range(len(seats[y])):
            status = seats[y][x]
            n = neighbor_counts[y][x]
            if status == 'L' and n == 0:
                status = '#'
            elif status == '#' and n >= max_occupied:
                status = 'L'

            new_seats[y][x] = status
            num_changed += status != seats[y][x]
            total_occupied += status == '#'
    return (num_changed, total_occupied)


def iter_seats(seats, neighbors_fn, max_occupied):
    s1 = [line.copy() for line in seats]
    s2 = [line.copy() for line in seats]
    s = [s1, s2]

    idx = 0
    while True:
        (changed, occ) = step(s[idx], s[1 - idx], neighbors_fn, max_occupied)
        if changed == 0:
            return occ
        idx = 1 - idx


def solution():
    seats = [list(line.strip())
             for line in open('../data/11.txt', 'r').readlines()]
    print(f'Answer 1: {iter_seats(seats, get_neighbor_counts, 4)}')
    print(f'Answer 2: {iter_seats(seats, get_visible_neighbor_counts, 5)}')

File: rust/src/day11.rs
use crate::common;
use std::str::FromStr;

#[derive(Debug, PartialEq, Copy, Clone)]
enum Seat {
  Floor,
  Vacant,
  Busy,
}

type SeatPlan = Vec<Vec<Seat>>;

impl FromStr for Seat {
  type Err = ();

  fn from_str(s: &str) -> Result<Self, Self::Err> {
    match s {
      "." => Ok(Seat::Floor),
      "L" => Ok(Seat::Vacant),
      "#" => Ok(Seat::Busy),
      _ => Err(()),
    }
  }
}

const OFFS: [(i32, i32); 8] = [
  (-1, -1),
  (0, -1),
  (1, -1),
  (-1, 0),
  (1, 0),
  (-1, 1),
  (0, 1),
  (1, 1),
];

fn get_seat(seats: &SeatPlan, (x, y): (i32, i32)) -> Option<Seat> {
  let (h, w) = (seats.len() as i32, seats[0].len() as i32);
  if x >= 0 && x < w && y >= 0 && y < h {
    Some(seats[y as usize][x as usize])
  } else {
    None
  }
}

fn get_neighbor_counts(seats: &SeatPlan) -> Vec<Vec<u32>> {
  let (h, w) = (seats.len() as i32, seats[0].len() as i32);
  let mut res = vec![vec![0; w as usize]; h as usize];
  for y in 0..h {
    for x in 0..w {
      res[y as usize][x as usize] = OFFS
        .iter()
        .filter_map(|(dx, dy)| get_seat(&seats, (x + dx, y + dy)))
        .map(|s| (s == Seat::Busy) as u32)
        .sum();
    }
  }
  res
}

fn get_visible_neighbor_counts(seats: &SeatPlan) -> Vec<Vec<u32>> {
  let (h, w) = (seats.len() as usize, seats[0].len() as usize);
  let mut res = vec![vec![0; w]; h];
  let mut can_see = vec![vec![false; w]; h];

  for (dx, dy) in OFFS.iter() {
    for j in 0..h {
      for i in 0..w {
        let x: usize = if *dx <= 0 { i } else { w - i - 1 };
        let y: usize = if *dy <= 0 { j } else { h - j - 1 };
        let pos = (x as i32 + dx, y as i32 + dy);
        let visible = match get_seat(&seats, pos) {
          Some(Seat::Floor) => can_see[pos.1 as usize][pos.0 as usize],
          Some(Seat::Busy) => true,
          Some(Seat::Vacant) => false,
          _ => false,
        };
        can_see[y][x] = visible;
        res[y][x] += visible as u32;
      }
    }
  }
  res
}

fn step(
  seats: &SeatPlan,
  new_seats: &mut SeatPlan,
  neighbors_fn: fn(&SeatPlan) -> Vec<Vec<u32>>,
  max_occupied: u32,
) -> (u32, u32) {
  let (h, w) = (seats.len() as usize, seats[0].len() as usize);
  let neighbor_counts = neighbors_fn(seats);
  let mut num_changed = 0;
  let mut total_occupied = 0;
  for y in 0..h {
    for x in 0..w {
      let n = neighbor_counts[y][x];
      let status = match seats[y][x] {
        Seat::Vacant if n == 0 => Seat::Busy,
        Seat::Busy if n >= max_occupied => Seat::Vacant,
        val => val,
      };
      new_seats[y][x] = status;
      num_changed += (status != seats[y][x]) as u32;
      total_occupied += (status == Seat::Busy) as u32;
    }
  }
  (num_changed, total_occupied)
}

fn iter_seats(
  seats: &SeatPlan,
  neighbors_fn: fn(&SeatPlan) -> Vec<Vec<u32>>,
  max_occupied: u32,
) -> u32 {
  let mut s1 = seats.clone();
  let mut s2 = seats.clone();
  let mut flip = true;
  loop {
    let (r1, r2) = if flip { (&s1, &mut s2) } else { (&s2, &mut s1) };
    let (changed, occupied) = step(r1, r2, neighbors_fn, max_occupied);
    if changed == 0 {
      return occupied;
    }
    flip = !flip;
  }
}

pub(crate) fn solution() {
  let seats: SeatPlan = common::read_lines(&common::data_file(11))
    .unwrap()
    .iter()
    .map(|line| {
      line
        .chars()
        .map(|c| c.to_string().parse().unwrap())
        .collect()
    })
    .collect();
  println!("Answer 1: {}", iter_seats(&seats, get_neighbor_counts, 4));
  println!(
    "Answer 2: {}",
    iter_seats(&seats, get_visible_neighbor_counts, 5)
  );
}
